File: Problem2.py
#Keeping 2 columns skip and take and adding optimal solution until that row
# TC: O(n)
# SC: O(n)
# Yes, this worked in leetcode

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        n = len(nums)
        skip = 0
        take = nums[0]
        for i in range(1, n):
            temp = skip
            skip = max(skip, take)
            take = temp + nums[i]
        return max(skip, take)

File: Problem_2.cpp
// Author: Subhed Chavan
// Batch: December 24
// Problem Statement: 198. House Robber
// Approach: Dynamic Programming
// Time Complexity: O(n)

class Solution
{
public:
    int rob(vector<int> &nums)
    {

        // Base Case
        if (nums.empty() || nums.size() == 0)
        {
            return 0;
        }

        int n = nums.size();
        int skip = 0;
        int take = 0;

        // Iterate over houses
        for (int i = 0; i < n; i++)
        {

            // Temporary store for Skip
            int tempSkip = skip;

            // Case 0
            skip = max(skip, take);

            // Case 1
            take = tempSkip + nums[i];
        }

        // Return the max of skip or take
        return max(skip, take);
    }
};